cpp/include/raft/util/reduction.cuh

```diff
@@ -0,0 +1,278 @@
+ *       function together. cub also uses too many registers
+ */
+template <typename T, typename ReduceLambda = raft::add_op>
+DI T blockReduce(T val, char* smem, ReduceLambda reduce_op = raft::add_op{})
+{
+  auto* sTemp = reinterpret_cast<T*>(smem);
+  int nWarps  = (blockDim.x + WarpSize - 1) / WarpSize;
+  int lid     = laneId();
+  int wid     = threadIdx.x / WarpSize;
+  val         = warpReduce(val, reduce_op);
+  if (lid == 0) sTemp[wid] = val;
+  __syncthreads();
+  val = lid < nWarps ? sTemp[lid] : T(0);
+  return warpReduce(val, reduce_op);
+}
+
+/**
+ * @brief warp-level weighted selection of an index.
+ * It selects an index with the given discrete probability
+ * distribution(represented by weights of each index)
+ * @todo benchmark whether a scan and then selecting within the ranges is more efficient.
+ * @param rng random number generator, must have next_u32() function
+ * @param weight weight of the rank/index.
+ * @param idx index to be used as rank
+ * @return only the thread0 will contain valid reduced result
+ */
+template <typename T, typename rng_t, typename i_t = int>
+DI T warpWeightedSelect(rng_t& rng, T& weight, i_t& idx)
+{
+  static_assert(std::is_integral<T>::value, "The type T must be an integral type.");
+#pragma unroll
+  for (i_t offset = raft::WarpSize / 2; offset > 0; offset /= 2) {
+    T tmp_weight = shfl(weight, laneId() + offset);
+    i_t tmp_idx  = shfl(idx, laneId() + offset);
+    T sum        = (tmp_weight + weight);
+    weight       = sum;
+    if (sum != 0) {
+      i_t rnd_number = (rng.next_u32() % sum);
+      if (rnd_number < tmp_weight) { idx = tmp_idx; }
+    }
+  }
+}
+
+/**
+ * @brief 1-D block-level weighted selection of an index.
+ * It selects an index with the given discrete probability
+ * distribution(represented by weights of each index)
+ *
+ * Let w_i be the weight stored on thread i.  We calculate the cumulative distribution function
+ * F_i = sum_{k=0..i} weight_i.
+ * Sequentially, we could select one of the elements with with the desired probability using the
+ * following method. We can consider that each element has a subinterval assigned: [F_{i-1}, F_i).
+ * We generate a uniform random number in the [0, F_i) range, and check which subinterval it falls.
+ * We return idx corresponding to the selected subinterval.
+ * In parallel, we do a tree reduction and make a selection at every step when we combine two
+ * values.
+ * @param rng random number generator, must have next_u32() function
+ * @param shbuf shared memory region needed for storing intermediate results. It
+ *             must alteast be of size: `(sizeof(T) + sizeof(i_t)) * WarpSize`
+ * @param weight weight of the rank/index.
+ * @param idx index to be used as rank
+ * @return only the thread0 will contain valid reduced result
+ */
+template <typename T, typename rng_t, typename i_t = int>
+DI i_t blockWeightedSelect(rng_t rng, T* shbuf, T weight = 1, i_t idx = threadIdx.x)
```

Comment:
@cjnolet I feel this (and `warpWeightedSelect`) shall belong to the `random` namespace. What is your opinion?

Reply:
I do really like the idea of moving the function over to the random namespace but still looking for ideas on how to best expose device functions through the public api directly. What do you guys think? 

We could do something like raft::random::device maybe? And then put it in a file suffixed with _device? Any other suggestions?

Reply:
`raft::random::device` namespace sounds fine to me. My first thought was to put the file in a device subfolder, but maybe it is better to keep it  closer to other sampling methods.

Reply:
I'm on board with the device sub-folder too, to better match the namespace, but either way works.